### ISSUE TITLE
Inference is slower when the input dimension of fully fused mlp is larger than 16

### ISSUE THREAD

**Issue by nogu-atsu** [NVlabs/tiny-cuda-nn#91]
I compared the speeds of 4-layer fully fused mlp with the hidden dimension of 64 and pure pytorch mlp using the PyTorch extension. The results show that fully fused mlp suddenly slows down when the input dimension exceeds 16, and pytorch mlp is faster when the input dimension exceeds 64.
Is this due to python overhead? Is there a workaround?

Here is the code I used (tested with A100 40GB GPU, cuda11.1, torch=1.10.1, ubuntu20.04, cmake3.23.1)
```
import tinycudann as tcnn
import torch
from torch import nn
import time
from torch.cuda.amp import autocast


def run(n_input_dims):
    print("n_input_dims:", n_input_dims)
    network_config = {
        "otype": "FullyFusedMLP",
        "activation": "ReLU",
        "output_activation": "None",
        "n_neurons": 64,
        "n_hidden_layers": 2
    }

    n_output_dims = 4
    n_neurons = network_config["n_neurons"]
    n_hidden_layers = network_config["n_hidden_layers"]

    # define tiny-cuda-nn model
    ffmlp = tcnn.Network(n_input_dims, n_output_dims, network_config)

    # define pytorch mlp
    layers = [nn.Linear(n_input_dims, n_neurons), nn.ReLU()]
    for i in range(n_hidden_layers):
        layers.append(nn.Linear(n_neurons, n_neurons))
        layers.append(nn.ReLU())
    layers.append(nn.Linear(n_neurons, n_output_dims))
    mlp = nn.Sequential(*layers).to("cuda")

    ffmlp.eval()
    mlp.eval()

    # measure time
    n_loop = 100

    x = torch.randn(2 ** 21, n_input_dims, device="cuda")

    with torch.set_grad_enabled(False):
        torch.cuda.synchronize()
        start = time.time()
        for i in range(n_loop):
            ffmlp_out = ffmlp(x)
        torch.cuda.synchronize()
        end = time.time()
        print(f"ffmlp time:", end - start)

        with autocast():
            torch.cuda.synchronize()
            start = time.time()
            for i in range(n_loop):
                mlp_out = mlp(x)
            torch.cuda.synchronize()
            end = time.time()
            print(f"torch mlp time:", end - start)


if __name__ == "__main__":
    run(n_input_dims=8)
    run(n_input_dims=16)
    run(n_input_dims=17)
    run(n_input_dims=32)
    run(n_input_dims=64)
```

Result
```
n_input_dims: 8
ffmlp time: 0.11553645133972168
torch mlp time: 0.5887284278869629
n_input_dims: 16
ffmlp time: 0.09747195243835449
torch mlp time: 0.5565247535705566
n_input_dims: 17
ffmlp time: 0.2910737991333008
torch mlp time: 0.6106259822845459
n_input_dims: 32
ffmlp time: 0.4063229560852051
torch mlp time: 0.5739290714263916
n_input_dims: 64
ffmlp time: 0.7818570137023926
torch mlp time: 0.6197981834411621
```

**Comment by Tom94** [comment]
Thank you very much for reporting this with a self-contained reproduction script!

Looks like there was indeed a performance regression. (To be precise: in `IdentityEncoding` -- `tcnn.Network` is actually a `tcnn.NetworkWithInputEncoding` with an identity `tcnn.Encoding()` upfront to do the padding and casting to fp16.)

Latest `master` should now be fixed. Here's the performance I observe on my 3090:
```
n_input_dims: 8
ffmlp time: 0.0920567512512207
torch mlp time: 0.6408650875091553
n_input_dims: 16
ffmlp time: 0.07889652252197266
torch mlp time: 0.6513733863830566
n_input_dims: 17
ffmlp time: 0.10257172584533691
torch mlp time: 0.663862943649292
n_input_dims: 32
ffmlp time: 0.11114835739135742
torch mlp time: 0.6861395835876465
n_input_dims: 64
ffmlp time: 0.15269994735717773
torch mlp time: 0.7538776397705078
```

There are still breakpoints at multiples of 16 due to extra data having to be loaded, but they're much more benign.
